# Introducción a las herramientas del cómputo científico.

Sergio A. Alcalá Corona <br>
Sergio A. Sánchez Chávez

---
---

# SISTEMA OPERATIVO LINUX

## Práctica 2

### Trabajo con Archivos (redireccionamiento, búsqueda, filtros y permisos)

Para esta práctica vamos a trabajar con una base de datos generada artificialmente. Esta se encuentra comprimida en el archivo **base.tar.gz**. 

---


## Antes de empezar ...

La **Base de Datos** esta organizada de la siguiente manera: 

Un directorio llamado _BaseDatos_ con subdirectorios organizados por alcaldías de la CDMX y sectores. 
Dentro de cada sector se encuentran archivos cuyo nombre esta dado por una fecha y dentro de cada uno hay una lista con 40 nombres y telefonos (todos generados aleatoriamente). 

Para descomprimir la base de datos utiliza el comando: 

`tar -xzf base.tar.gz`

Para esta práctica será de ayuda utilizar el comando _**`wc`**_ (word counting)

`wc -l archivo` para contar líneas

`wc -w archivo` para contar palabras

Si necesitas más información de este comando utiliza el manual

## Indicaciones

- Tendrás que abrir tu _Jupyterlab_ y realizarás los ejercicios en un notebook de Bash.

- Sólo es válido usar comandos de Bash (no Python, ni algún otro lenguaje de programación)

- Todos los comandos que uses (a menos que se indique lo contrario) deben de estar en el notebook.

- __*TRATA DE USAR EL MENOR NUMERO DE COMANDOS POSIBLES.*__

- Los archivos resultantes deben de colocarse en un directorio llamado __*Resultado*__. Esta carpeta la deberás de comprimir y subirla al Classroom. El archivo que debes subir tiene que estar comprimido en un archivo .tar.gz para ello utilizarás el comando:

`tar -czfv resultado.tar.gz Resultado/`

## cat,  head, tail

In [1]:
tar -xzf base.tar.gz

**1.** Copia el archivo (utilizando comandos de bash) que se encuentra en `directorio/TLALPAN/sectorH/names_2009-08-30.txt` a la carpeta `Resultado`.  

Ahora colócate en el directorio `Resultado` y ejecuta los siguientes comandos, para responder a las siguientes preguntas:  
    

In [2]:
mkdir Resultado && cp directorio/TLALPAN/sectorH/names_2009-08-30.txt ./Resultado && cd Resultado

**2.** ¿Cuál es el resultado de ejecutar los siguientes comandos en la terminal? 

$\; \; \; \; \;$ a) `cat < names_2009-08-30.txt > names_cat.txt`
    
$\; \; \; \; \;$ b) `cat < names_2009-08-30.txt >> names_cat.txt`
    
$\; \; \; \; \;$ c) ¿Cuál es la diferencia entre ambos? Explícalo con tus propias palabras. 

In [3]:
cat < names_2009-08-30.txt > names_cat.txt && cat < names_2009-08-30.txt >> names_cat.txt

a) El resultado es el darle de entrada el archivo `names_2009-08-30.txt` a `cat` y que creara el archivo `names_cat.txt`.

b) El resultado es el darle de entrada el archivo `names_2009-08-30.txt` a `cat` y que concatenara el contenido al archivo `names_cat.txt`.

c) La diferencia de los dos comandos es el uso entre `>` y `>>`, el primero sobrescribe si ya existe un archivo y el otro concatena al final del archivo si es que ya existe.

**3.** ¿Cuál es la diferencia entre estos comandos?
 
$\; \; \; \; \;$ a) `cat names.txt 2> err.dat`
    
$\; \; \; \; \;$ b) `cata names.txt 2> err.dat`
    

In [4]:
cat names.txt 2> err.dat
cata names.txt 2> err.dat

: 127

Que uno usa el comando `cat` y el otro usa en comando `cata`, que no existe y dará error. Ambos redireccionan el standerr al output por lo que si producen un archivo de salida, pero con contenidos diferentes.

**4.** Ejecuta: `cat > feliz` enseguida escibre algo y luego ejecuta `Ctrl+D` dos veces. ¿qué fue lo que sucedio?

`cat > feliz`

`hola mundo`

`Ctrl+D`

`Ctrl+D`

Lo que sucedió fue que el texto que introducimos después de ejecutar el comando fue introducido en un nuevo archivo llamado `feliz`. Y con `Ctrl+D` salimos de la instancia del programa, y al presionarlo una vez más, salimos de la terminal.

**5.** ¿Existe algúna forma, con los comandos de bash, de mostrar el archivo de forma inversa? En caso de existir, utilízalo para ver el archivo `directorio/COYOACAN/sectorG/names_2005-09-27.txt` y guarda la salida en el archivo `catinverso.dat` 

Existe la opción de usar `tail -r` en algunas distros, en la mía no, así que usaré `tac` que es similar a `cat` pero para leer de forma inversa. Tampoco existe el archivo `names_2005-09-27.txt` en el directorio, así que usaré el archivo `names_2005-09-27.dat` asumiendo que en realidad se refiere a ese.

In [5]:
tac ../directorio/COYOACAN/sectorG/names_2005-09-27.dat > ./catinverso.dat

**6.** Crea un solo archivo con el nombre `todosarchivos.txt` que tenga el contenido de todos los archivos del directorio `directorio/CUAUHTEMOC/sectorE/` 

In [6]:
cat ../directorio/CUAUHTEMOC/sectorE/* > todosarchivos.txt

**7.** Coloca en un archivo con el nombre `colacabeza` las primeras 23 líneas numeradas y las últimas 32 líneas numeradas del archivo `todosarchivos.txt`

In [7]:
cat -n todosarchivos.txt | head -23 > colacabeza && cat -n todosarchivos.txt | tail -32 >> colacabeza

**Opcional**: Numera las líneas.  Genera un archivo donde se indiquen con numeros las líneas de este archivo.

In [8]:
cat -n colacabeza > colacabezanumerado

## Procesos y Permisos

**8.** ¿Cuál es la diferencia entre los comandos `top`, `htop` y `ps`?

Con ps podemos ver los procesos que en el momento de la ejecución del comando se estaban ejecutando, con top podemos ver los procesos que se están ejecutando de manera dinámica, y con htop (human top) tenemos más opciones para visualizar en tiempo real la ejecución de procesos.

**9**. Crea un archivo llamado `MisProcesos.prs` que contenga todos los procesos de tu usuario activos en tu equipo, junto con sus características.

In [9]:
ps -fea > MisProcesos.prs

**10.** Explica que hace el comando `umask` y pon un ejemplo de su uso. 

El comando `umask` determina los permisos por defecto que se le darán a los nuevos archivos y directorios que sean 
creados en el futuro. Los permisos de los nuevos archivos serán calculados de acuerdo a los permisos por defecto
    máximos de un directorio y un archivo (777 y 666 respectivamente) y se les serán sustraídos los valores que definamos en el comando `umask`, a excepción de permisos de ejecución que esos en el caso de los archivos tendrán que ser agregados y no serán otorgados. 
    
Este comando se usa así:

`$ umask 453`

**11.** Copia la carpeta `CUAUHTEMOC` dentro de `Resultados/` y cambia los permisos de todos los archivos de la siguiente manera sólo lectura y escritura para el dueño, sólo ejecución para el grupo y sólo lectura para los demás.

In [10]:
cp -r ../directorio/CUAUHTEMOC . && find CUAUHTEMOC/ -type f -exec chmod 614 {} +

**12.** Cambia los permisos (por el método octal) de todos los archivos con terminación __.txt__ dentro de la carpeta `CUAUHTEMOC/sectorG/`, de tal manera que el dueño tenga lectura, escritura y ejecución, sólo lectura para el grupo y ningún permiso para los demás.

In [11]:
find CUAUHTEMOC/sectorG/ -name "*.txt" -exec chmod 740 {} +

**13.** Cambia los permisos de todos los archivos con terminación __.dat__ dentro de la carpeta `CUAUHTEMOC/sectorP/`, de tal manera que el dueño tenga todos los permisos, sólo ejecución para el grupo y sólo lectura  para los demás. Aplica un `ls -l` para ver el resultado y guarda eso en un archivo con nombre `permisos.dat`

In [12]:
find CUAUHTEMOC/sectorP/ -name "*.txt" -exec chmod 714 {} + && ls -l CUAUHTEMOC/sectorP/ > permisos.dat

**14.** Utiliza el método (númerico o simbólico) que quieras para cambiar los permisos de todos los archivos en cuyo nombre este la fecha de mayo (05) que están dentro de la carpetas `CUAUHTEMOC/sectorP/` y `CUAUHTEMOC/sectorO/`. Los permisos deberán quedar de la siguiente manera todos los permisos para el dueño y ningún permiso para el grupo y los demás usuarios.

In [13]:
find CUAUHTEMOC/sectorP/ CUAUHTEMOC/sectorO/ -type f -name "*_*-05-*.*" -exec chmod 700 {} +

## find y grep

**15.** ¿Cuántos archivos tienen una extensión de `.txt` y cuántos tienen una extension `.dat`, en toda la base de datos?

In [14]:
find ../directorio/ -name "*.txt" | wc -l && find ../directorio/ -name "*.dat" | wc -l

2036
2121


Son 2036 archivos con extensión `.txt` y 2121 archivos con la extensión `.dat` en el directorio.

**16.** ¿Cuántos veces aparece _tu nombre_ (sin apellido) en todos los archivos? 

In [15]:
grep -r "Carlos" ../directorio/ | wc -l

503


Mi nombre `"Carlos"` aparece 503 veces en la carpeta `directorio`.

**17**. Realiza lo siguiente:

$\; \; \; \; \;$ a) Crea un archivo `cuauhtemoc.data`  que contenga las rutas de todos los archivos del directorio `CUAUHTEMOC` (incluyendo sus subdirectorios) cuya extensión sea `.dat`

In [16]:
find ../directorio/CUAUHTEMOC/ -name "*.dat" > cuauhtemoc.data

$\; \; \; \; \;$ b) Crea un archivo coyoacan.dir  que contenga las rutas de todos los archivos del directorio `COYOACAN` (incluyendo sus subdirectorios) que en su interior contienen la palabra "Perez". 

In [17]:
grep -lr "Perez" ../directorio/COYOACAN/ > coyoacan.dir

$\; \; \; \; \;$ c) Agrega a ese archivo las rutas de los archivos del directorio `VENUSTIANO_CARRANZA` cuya extensión sea `.txt` y que en su interior contengan la palabra "Luis".

In [18]:
find ../directorio/VENUSTIANO_CARRANZA/ -name "*.txt" -exec grep -lr "Luis" {} + >> coyoacan.dir

$\; \; \; \; \;$ d) Crea un archivo llamado `MiUsuario.usr` que contenga las rutas de todos los archivos de la base (incluyendo sus subdirectorios) que en su interior contengan tu nombre.

In [19]:
grep -lr "Carlos" ../directorio/ > MiUsuario.usr

**18**. Contesta lo siguiente:

$\; \; \; \; \;$ a) ¿En cuántos archivos __NO__ aparecen tus nombres?

In [20]:
grep -vlr "Carlos" ../directorio/ | wc -l

4157


Mi nombre no aparece en 4157 archivos dentro de la base de datos.

$\; \; \; \; \;$ b) ¿Cuántos veces coincide tus nombres con un archivo que tenga el año de tu nacimiento?

In [21]:
find ../directorio/ -name "*_2003-*" -exec grep "Carlos\|Emilio" {} + | wc -l

55


Mis nombres coinciden con archivos que tengan mi fecha de nacimiento 55 veces.

$\; \; \; \; \;$ c) ¿De cuántos directorios y subdirectorios está compuesta la base de datos?

In [22]:
find ../directorio/ -type d | wc -l

187


La base de datos cuenta con 187 directorios y subdirectorios.

**19.** Realiza lo siguiente:

$\; \; \; \; \;$ a) Encuentra todos los archivos que cuenten con todos los permisos para todos los usuarios en el directorio `BENITO_JUAREZ` y escribelos en archivo llamado `juarez.prm`

In [23]:
find ../directorio/BENITO_JUAREZ/ -type f -perm 777 > juarez.prm

$\; \; \; \; \;$ b) Encuentra todos los archivos que tengan la extensión `.tar` en el directorio `IZTAPALAPA` y escribelos en archivo llamado `iztapalpall.compress`

In [24]:
find ../directorio/IZTAPALAPA/ -type f -name "*.tar" > iztapalpall.compress

**20.** Contesta lo siguiente:

$\; \; \; \; \;$ a) ¿Cuántos archivos tienen un tamaño menor a 1150 bytes?

In [25]:
find ../directorio/ -type f -size -1150c | wc -l

4157


Hay 4157 archivos con un tamaño menor a 1150 bytes en la base de datos.

$\; \; \; \; \;$ b) ¿Cuántos archivos coinciden con el año 2020 en su nombre y además pertencen al `sectorS`?

In [26]:
find ../directorio/*/sectorS -name "*_2020-*" | wc -l

8


Hay 8 archivos que coinciden con el año 2020 y pertenecen al sectorS.

In [30]:
cd .. && tar -czf - Resultado/ | gzip > resultado.tar.gz